### PR TITLE
[Placement Group] Leasing context refactoring part 2

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This refactors several in-memory index regarding leasing status into `LeasingContext`. This will be useful for implementing atomic placement group creation because it will be involved in more states and in-memory indexes.

Note
- Write unit tests
- Add docstring to interface

are skipped in this PR because the next PR (atomic creation PR) will anyway change them a lot.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc [hunk 1]
```diff
@@ -48,6 +48,11 @@ std::vector<std::shared_ptr<BundleSpecification>> GcsPlacementGroup::GetBundles(
   return ret_bundles;
 }
 
+size_t GcsPlacementGroup::GetBundleSize() const {
+  const auto &bundles = placement_group_table_data_.bundles();
+  return bundles.size();
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:53]:
```suggestion
  return placement_group_table_data_.bundles().size();
```

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc [hunk 2]
```diff
@@ -214,46 +214,26 @@ void GcsPlacementGroupScheduler::ScheduleUnplacedBundles(
...
-
-          if (++(*finished_count) == bundles.size()) {
-            OnAllBundleSchedulingRequestReturned(placement_group, bundles,
-                                                 bundle_locations, failure_callback,
-                                                 success_callback);
-          }
-        });
+    leasing_context->MarkLeaseStarted(node_id, bundle);
+
+    ReserveResourceFromNode(bundle, gcs_node_manager_.GetNode(node_id),
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:228]: @rkooo567 I'm not sure if callback will be called if raylet is dead, do you know the logic of this?

**Review by ffbin — APPROVED** [review]
LGTM

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:53]: This method will be removed. It is unnecessary for now.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:228]: I think it wont. Are you suggesting to have a callback map and empty them when the node is dead or placement group is removed?

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:53]: I will remove this method. We don't need it for now.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:228]: I will address this in the next PR. Added a comment for now.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:228]: Added a comment for now.

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h [hunk 3]
```diff
@@ -75,6 +75,9 @@ class GcsPlacementGroup {
   /// Get the bundles of this placement_group (including unplaced).
   std::vector<std::shared_ptr<BundleSpecification>> GetBundles() const;
 
+  /// Get the size of bundles (including unplaced).
+  size_t GetBundleSize() const;
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.h:79]: What does size mean? NumBundlesTotal()? Also, should not be size_t.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 4]
```diff
@@ -131,17 +131,68 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  ALL_RETURNED,
+  /// Placement group has been removed, and this leasing is not valid.
+  CANCELLED
+};
+
 /// A data structure that encapsulates information regarding bundle resource leasing
 /// status.
 class LeasingContext {
-  // TODO(sang): Implement in the next PR.
+ public:
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:147]: Consider renaming this to `LeaseStatusTracker`. Context is a pretty vague term

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h [hunk 5]
```diff
@@ -131,17 +131,68 @@ class GcsStrictSpreadStrategy : public GcsScheduleStrategy {
...
+  LeasingContext(std::shared_ptr<GcsPlacementGroup> placement_group,
+                 std::vector<std::shared_ptr<BundleSpecification>> &unplaced_bundles);
+  ~LeasingContext() = default;
+
+  bool MarkLeaseStarted(const ClientID &node_id,
+                        std::shared_ptr<BundleSpecification> bundle);
+  void MarkLeaseReturned(const ClientID &node_id,
+                         std::shared_ptr<BundleSpecification> bundle,
+                         const Status &status);
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:156]: Please document the public methods.

**Review by ericl — CHANGES_REQUESTED** [review]

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h (hunk 3 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.h]: This method is removed.

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h (hunk 5 above)
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:156]: Would you mind me doing this in the next PR? Most of signature will be changed in the atomic creation PR (I wrote this in the PR description). I can still do it if you'd want me to, so just lmk.

**Review by ericl — APPROVED** [review]
Sounds good

**Pull request merged**